cmd/security-profiles-operator-init/main.go

```diff
@@ -0,0 +1,233 @@
+*/
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	"github.com/urfave/cli/v2"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2/klogr"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"sigs.k8s.io/security-profiles-operator/api/v1alpha1"
+	"sigs.k8s.io/security-profiles-operator/internal/pkg/config"
+	"sigs.k8s.io/security-profiles-operator/internal/pkg/version"
+)
+
+const (
+	jsonFlag                 = "json"
+	operatorNamespace        = "security-profiles-operator"
+	resourceName             = "seccompprofiles"
+	kubeletSeccompRootEnvKey = "KUBELET_SECCOMP_ROOT"
+	operatorRootEnvKey       = "OPERATOR_ROOT"
+	operatorLinkEnvKey       = "OPERATOR_SYMLINK"
+	operatorUID              = 2000
+	operatorGID              = 2000
+)
+
+var setupLog = ctrl.Log.WithName("setup")
+
+func main() {
+	ctrl.SetLogger(klogr.New())
+
+	app := cli.NewApp()
+	app.Name = config.OperatorName + "-init"
+	app.Usage = "Kubernetes Seccomp Operator Setup"
+	app.Description = "Setup script for the Kubernetes Seccomp Operator."
+	app.Version = version.Get().Version
+	app.Action = run
+	app.Commands = cli.Commands{
+		&cli.Command{
+			Name:    "version",
+			Aliases: []string{"v"},
+			Usage:   "display detailed version information",
+			Flags: []cli.Flag{
+				&cli.BoolFlag{
+					Name:    jsonFlag,
+					Aliases: []string{"j"},
+					Usage:   "print JSON instead of text",
+				},
+			},
+			Action: func(c *cli.Context) error {
+				v := version.Get()
+				res := v.String()
+				if c.Bool(jsonFlag) {
+					j, err := v.JSONString()
+					if err != nil {
+						return errors.Wrap(err, "unable to generate JSON from version info")
+					}
+					res = j
+				}
+				print(res)
+				return nil
+			},
+		},
+	}
+	app.Flags = []cli.Flag{
+		&cli.StringFlag{
+			Name:  "path",
+			Usage: "mount path for default profiles to bootstrap",
+		},
+	}
+
+	if err := app.Run(os.Args); err != nil {
+		setupLog.Error(err, "running security-profiles-operator")
+		os.Exit(1)
+	}
+}
+
+func run(c *cli.Context) error {
+	v := version.Get()
+	setupLog.Info(
+		"starting security-profiles-operator-init",
+		"version", v.Version,
+		"gitCommit", v.GitCommit,
+		"gitTreeState", v.GitTreeState,
+		"buildDate", v.BuildDate,
+		"goVersion", v.GoVersion,
+		"compiler", v.Compiler,
+		"platform", v.Platform,
+	)
+
+	if err := v1alpha1.AddToScheme(scheme.Scheme); err != nil {
+		return errors.Wrap(err, "setting up client")
+	}
+
+	clientset, err := getClient()
+	if err != nil {
+		return errors.Wrap(err, "getting REST client")
+	}
+
+	profiles, err := getProfiles(c.String("path"))
+	if err != nil {
+		return errors.Wrap(err, "getting profiles")
+	}
+
+	// Create SeccompProfile resources
+	ctx := context.Background()
+	for name, sp := range profiles {
```

Comment:
Is the main reason for maintaining our own init container such that all of the "bootstrap" profiles will be reflected as CRDs in the cluster as well? I think we can modify deployment such that instances of the CRD are not created until the CRD itself is. For instance, helm handles this with a dedicated `crds` directory. Not strictly against maintaining our own init image, but as has been mentioned before, it is a larger attack surface.

Reply:
The main reason is item number 2 in [this commit message](https://github.com/kubernetes-sigs/security-profiles-operator/pull/138/commits/92481d97f21d7c978bb6f329ed15eab09b7309ea) which is that the controller container itself cannot start if the seccomp profile for the container image itself doesn't exist on the host, so the controller can't reconcile the `SeccompProfile` because it hasn't started yet. I'm certainly open to other ideas here.

Reply:
@cmurphy We currently resolve that chicken-egg situation by not relying on any of the operator's functionality to deploy its profile, we simply use the built-in "ConfigMap to Volume" mount and copy the profile with `cp`, that's why that specific `ConfigMap` does not have the `seccomp.security.kubernetes.io/profile: "true"`. Here's where we do it:

https://github.com/kubernetes-sigs/security-profiles-operator/blob/49f5464dc2e05fbfaa9d176ca0afb156034824a8/deploy/base/operator.yaml#L43

As @hasheddan mentioned, we took that approach to limit the attack surface running without any profile-based enforcement. Do you think we can get away without changing that?